text/0000-default-field-values.md

```diff
@@ -0,0 +1,1979 @@
+  performance characteristics.
+
++ Keeping default values limited to `const` expressions ensures that if
+  the following situation develops:
+
+  ```rust
+  // Crate A:
+  pub struct Foo {
+      bar: u8 = const_expr,
+  }
+
+  // Crate B:
+  const fn baz() -> Foo {
+      Foo { .. }
+  }
+  ```
+
+  then crate A cannot suddenly, and unawares, cause a semver breakage
+  for crate B by replacing `const_expr` with `non_const_expr` since
+  the compiler would reject such a change (see lemmas 1-2).
+  Thus, enforcing constness gives a helping hand in respecting semantic version.
+
+  Note that if Rust would ever gain a mechanism to state that a
+  function will not diverge, e.g.:
+
+  ```rust
+  nopanic fn foo() -> u8 { 42 } // The weaker variant; more easily attainable.
+  total fn bar() -> u8 { 24 } // No divergence, period.
+  ```
+
+  then the same semver problem would manifest itself for those types of
+  functions. However, Rust does not have any such enforcement mechanism
+  right now and if it did, it is generally harder to ensure that a function
+  is total than it is to ensure that it is deterministic; thus, while
+  it is regrettable, this is an acceptable trade-off.
+
++ Finally, note that `const fn`s, when coupled with planned future extensions,
+  can become quite expressive. For example, it is already planned that `loop`s,
+  `match`es, `let` statements, and `panic!(..)`s will be allowed.
+  Another feasible extension in the future is allocation.
+
+  Therefore, constant expressions should be enough to satisfy most expressive
+  needs.
+
+## Instead of `Foo { ..Default::default() }`
+
+As an alternative to the proposed design is either explicitly writing out
+`..Default::default()` or extending the language such that `Foo { .. }` becomes
+sugar for `Foo { ..Default::default() }`. While the latter idea does not satisfy
+any of the [motivation] set out, the former does to a small extent.
+
+In particular, `Foo { .. }` as sugar slightly improves ergonomics.
+However, it has some notable problems:
+
++ Because it desugars to `Foo { ..Default::default() }`, it cannot be required
+  that the expression is a constant one. This carries all the problems noted in
+  the previous section on why default field values should be a `const` context.
+
++ There is no way of implementing a `Default` implementation that has mandatory
+  fields for users to specify during value construction.
+
++ It provides zero improvements to the ergonomics of *specifying* defaults,
+  only for using them. Arguably, the most important aspect of this RFC is
+  not the syntax `Foo { .. }` but rather the ability to provide default values
+  for field.
+
++ By extension, the improvement to documentation clarity is lost.
+
++ The trait `Default` must now become a `#[lang_item]`. This is a sign of
+  increasing the overall magic in the system; meanwhile, this proposal makes
+  the default values provided usable by other custom derive macros.
```

Comment:
Instead we're adding a new primitive expression syntax -- that also increases overall magic. In fact it increases overall magic *more* than a lang item, since a lang item at least still follow the already-known rules of existing items.

So IMO this is a bad argument, and should be removed from the list. Focus on the uncontroversial points. The point of an RFC is not to be a sales document that presents a proposal as the best thing since sliced bread, it is to fairly weigh the technical pros and cons.

Reply:
I just tried to accurately incorporate all of the arguments that I've encountered when discussing this feature. I'm ok with removing this, but would prefer to include it as it *has* been discussed and wouldn't want language historians to have to dig too much to determine whether it was not discussed or discussed and disregarded.

Reply:
I'd reduce the bullet to "this proposal makes the default values provided usable by other custom derive macros."

(Arguments that are based on the "overall magic" metric are sometimes worthwhile, but when it is questionable whether the overall magic is rising or falling, then its better to just stop using that metric entirely.)